### PR TITLE
Refs [TASK] [36129] Xử lí so sánh thông tin của hai hero

### PR DESCRIPTION
**Pull request by mithong**
## Related Tickets
- [#Task 36129: Content](https://edu-redmine.sun-asterisk.vn/issues/36129)
## WHAT
- hiển thị hai tướng và so sánh các tham số với nhau
## Evidence (Screenshot or Video)
image

## Review Checklist

Category | View Point | Description | Self review | Reviewer2 (name)
--- | --- | --- | --- | ---
Conventions | Sun* coding conventions | https://github.com/framgia/coding-standards/tree/master/eng/android |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Conventions | Kotlin coding conventions | https://kotlinlang.org/docs/coding-conventions.html |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Redmine | Sun* Redmine working process  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md |<li>- [ ] yes</li>|<li>- [ ] yes</li>  

## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/java/com/sun/hero_01/ui/champion/ChampionFragment.kt [hunk 1]
```diff
@@ -1,22 +1,64 @@
...
 import com.sun.hero_01.ui.search.SearchFragment
 import com.sun.hero_01.utils.OnItemRecyclerViewListener
 import com.sun.hero_01.utils.extensions.replaceFragment
 import kotlinx.android.synthetic.main.fragment_champion.*
+import kotlinx.android.synthetic.main.item_layout_hero.*
 
 class ChampionFragment : BaseFragment(), ChampionContract.View, OnItemRecyclerViewListener<Hero> {
 
-    private val adapter by lazy { ChampionAdapter(this) }
+    val shared = "kotlin shared"
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/hero_01/ui/champion/ChampionFragment.kt]: check convention

File: app/src/main/java/com/sun/hero_01/ui/champion/ChampionFragment.kt [hunk 2]
```diff
@@ -1,22 +1,64 @@
...
+        popupMenu.setOnMenuItemClickListener{
+            when(it.itemId){
+                R.id.menuCompare -> {
+                    onAddHero(its)
+                    true
+                }
+                else -> false
+            }
+        }
+        popupMenu.show()
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/hero_01/ui/champion/ChampionFragment.kt]: scope function

File: app/src/main/java/com/sun/hero_01/ui/champion/ChampionFragment.kt [hunk 3]
```diff
@@ -1,22 +1,64 @@
...
+    }
+
+    private fun onAddHero(its: String?) {
+        val sharedPreferences: SharedPreferences = context!!.getSharedPreferences(shared, Context.MODE_PRIVATE)
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/hero_01/ui/champion/ChampionFragment.kt]: check nullable

File: app/src/main/java/com/sun/hero_01/ui/champion/ChampionFragment.kt [hunk 4]
```diff
@@ -1,22 +1,64 @@
...
+        val editor = sharedPreferences.edit()
+        if(sharedPreferences.getString("First","").equals("")){
+            editor.putString("First",its)
+        }
+        else if(sharedPreferences.getString("First","").equals(its)){
+            Toast.makeText(context,resources.getString(R.string.cantcompare),Toast.LENGTH_LONG).show()
+        }
+        else{
+            editor.putString("Second",its)
+            replaceFragment(CompareFragment.newInstance(), R.id.frameContainer)
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/hero_01/ui/champion/ChampionFragment.kt]: hard code

File: app/src/main/java/com/sun/hero_01/ui/compare/CompareConstract.kt [hunk 5]
```diff
@@ -0,0 +1,16 @@
+package com.sun.hero_01.ui.compare
+
+import com.sun.hero_01.base.BasePresenter
+import com.sun.hero_01.data.model.HeroDetail
+
+interface CompareConstract {
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/hero_01/ui/compare/CompareConstract.kt]:
```suggestion
interface CompareContract {
```

File: app/src/main/java/com/sun/hero_01/ui/compare/CompareFragment.kt [hunk 6]
```diff
@@ -1,11 +1,134 @@
...
+import kotlin.math.roundToInt
 
-class CompareFragment {
+class CompareFragment : BaseFragment(), CompareConstract.View {
+
+    override val layoutResourceId = R.layout.fragment_compare
+    override var bottomNavigationViewVisibility = View.GONE
+
+    private var comparePresenter: ComparePresenter? = null
+    val shared = "kotlin shared"
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/hero_01/ui/compare/CompareFragment.kt]: check convention

File: app/src/main/java/com/sun/hero_01/ui/compare/CompareFragment.kt [hunk 7]
```diff
@@ -1,11 +1,134 @@
...
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        initView()
+    }
+
+    private fun initView() {
+        val sharedPreferences: SharedPreferences = context!!.getSharedPreferences(shared, Context.MODE_PRIVATE)
+        val firstHero = sharedPreferences.getString("First","")
+        val secondHero = sharedPreferences.getString("Second","")
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/hero_01/ui/compare/CompareFragment.kt]: as above

File: app/src/main/java/com/sun/hero_01/ui/compare/CompareFragment.kt [hunk 8]
```diff
@@ -1,11 +1,134 @@
...
+        }
+    }
+
+    private fun onShowProgressBar(progressBar: ProgressBar?, first: String, second: String) {
+        val firstNumber = first.toFloat()
+        val secondNumber = second.toFloat()
+        progressBar?.let {
+            if(firstNumber > secondNumber){
+                it.progressBackgroundTintList = ColorStateList.valueOf(Color.DKGRAY)
+            }else if(firstNumber < secondNumber){
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/hero_01/ui/compare/CompareFragment.kt]: reformat code
còn nếu case = thì sao e?

File: app/src/main/java/com/sun/hero_01/ui/compare/CompareFragment.kt [hunk 9]
```diff
@@ -1,11 +1,134 @@
...
+                it.progressTintList = ColorStateList.valueOf(Color.DKGRAY)
+            }
+            it.max = 100
+            it.progress = (firstNumber * 100 / (firstNumber + secondNumber)).roundToInt()
+            Log.d("aaaaaaaa",it.max.toString() +"/"+it.progress)
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/hero_01/ui/compare/CompareFragment.kt]: remove log

File: app/src/main/java/com/sun/hero_01/ui/compare/CompareFragment.kt [hunk 10]
```diff
@@ -1,11 +1,134 @@
...
+            Toast.makeText(requireActivity(), it.message, Toast.LENGTH_SHORT).show()
+        }
+    }
+
+    override fun onStop() {
+        super.onStop()
+        val sharedPreferences: SharedPreferences = context!!.getSharedPreferences(shared, Context.MODE_PRIVATE)
+        val editor = sharedPreferences.edit()
+        editor.remove("First")
+        editor.remove("Second")
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/hero_01/ui/compare/CompareFragment.kt]: hard code

File: app/src/main/java/com/sun/hero_01/ui/favorite/FavoriteFragment.kt [hunk 11]
```diff
@@ -60,6 +60,9 @@ class FavoriteFragment : BaseFragment(), FavoriteContact.View {
                         it.search(editSearchFavorite.text.toString())
                     }
                 }
+                else{
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/hero_01/ui/favorite/FavoriteFragment.kt]: reformat code

File: app/src/main/java/com/sun/hero_01/ui/compare/CompareFragment.kt (hunk 8 above)
Comment (mithong) [reply app/src/main/java/com/sun/hero_01/ui/compare/CompareFragment.kt]: case = e không thay đổi màu nên e không đưa vô ạ

File: app/src/main/java/com/sun/hero_01/ui/champion/ChampionFragment.kt [hunk 12]
```diff
@@ -1,22 +1,69 @@
...
 class ChampionFragment : BaseFragment(), ChampionContract.View, OnItemRecyclerViewListener<Hero> {
 
-    private val adapter by lazy { ChampionAdapter(this) }
+    private val adapterChampion by lazy {
+        ChampionAdapter(this) { idName, view ->
+            onShowMenu(idName, view)
+        }
+    }
+
+    private val sharedKey = "kotlin shared"
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/hero_01/ui/champion/ChampionFragment.kt]: https://github.com/framgia/coding-standards/blob/master/eng/android/codingstyleguide.md#210-string-constants-naming-and-values

File: app/src/main/java/com/sun/hero_01/ui/champion/ChampionFragment.kt [hunk 13]
```diff
@@ -1,22 +1,69 @@
...
+
+    private fun onAddHero(its: String?) {
+        val sharedPreferences: SharedPreferences? = context?.getSharedPreferences(sharedKey, Context.MODE_PRIVATE)
+        val editor = sharedPreferences?.edit()
+        val firstHero = resources.getString(R.string.first_hero)
+        val secondHero = resources.getString(R.string.second_hero)
+        if(sharedPreferences?.getString(firstHero,"").equals("")){
+            editor?.putString(firstHero, its)
+        }
+        else if(sharedPreferences?.getString(firstHero,"").equals(its)){
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/hero_01/ui/champion/ChampionFragment.kt]: reformat code

File: app/src/main/java/com/sun/hero_01/ui/champion/ChampionFragment.kt [hunk 14]
```diff
@@ -1,22 +1,69 @@
...
+            show()
+        }
+    }
+
+    private fun onAddHero(its: String?) {
+        val sharedPreferences: SharedPreferences? = context?.getSharedPreferences(sharedKey, Context.MODE_PRIVATE)
+        val editor = sharedPreferences?.edit()
+        val firstHero = resources.getString(R.string.first_hero)
+        val secondHero = resources.getString(R.string.second_hero)
+        if(sharedPreferences?.getString(firstHero,"").equals("")){
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/hero_01/ui/champion/ChampionFragment.kt]: reformat code

File: app/src/main/java/com/sun/hero_01/ui/champion/ChampionFragment.kt [hunk 15]
```diff
@@ -1,22 +1,69 @@
...
+            editor?.putString(firstHero, its)
+        }
+        else if(sharedPreferences?.getString(firstHero,"").equals(its)){
+            Toast.makeText(context,resources.getString(R.string.cantcompare),Toast.LENGTH_LONG).show()
+        }
+        else{
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/hero_01/ui/champion/ChampionFragment.kt]: reformat code

File: app/src/main/java/com/sun/hero_01/ui/compare/CompareFragment.kt [hunk 16]
```diff
@@ -1,11 +1,130 @@
...
+import kotlin.math.roundToInt
 
-class CompareFragment {
+class CompareFragment : BaseFragment(), CompareContract.View {
+
+    override val layoutResourceId = R.layout.fragment_compare
+    override var bottomNavigationViewVisibility = View.GONE
+
+    private var comparePresenter: ComparePresenter? = null
+    private val shared = "kotlin shared"
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/hero_01/ui/compare/CompareFragment.kt]: https://github.com/framgia/coding-standards/blob/master/eng/android/codingstyleguide.md#210-string-constants-naming-and-values

File: app/src/main/java/com/sun/hero_01/ui/compare/CompareFragment.kt [hunk 17]
```diff
@@ -1,11 +1,130 @@
...
+        }
+    }
+
+    override fun onStop() {
+        super.onStop()
+        val sharedPreferences: SharedPreferences? = context?.getSharedPreferences(shared, Context.MODE_PRIVATE)
+        val editor = sharedPreferences?.edit()
+        editor?.remove(resources.getString(R.string.first_hero))
+        editor?.remove(resources.getString(R.string.first_hero))
+        editor?.apply()
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/hero_01/ui/compare/CompareFragment.kt]: scope function

File: app/src/main/java/com/sun/hero_01/ui/champion/ChampionFragment.kt [hunk 18]
```diff
@@ -1,22 +1,67 @@
...
-    private val adapter by lazy { ChampionAdapter(this) }
+    private val adapterChampion by lazy {
+        ChampionAdapter(this) { idName, view ->
+            onShowMenu(idName, view)
+        }
+    }
+
+    private fun onShowMenu(its: String?, view: View?) {
+        val popupMenu = PopupMenu(context,view)
+        popupMenu.apply {
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/hero_01/ui/champion/ChampionFragment.kt]:
```suggestion
        PopupMenu(context,view).apply {
```

File: app/src/main/java/com/sun/hero_01/ui/champion/ChampionFragment.kt [hunk 19]
```diff
@@ -1,22 +1,67 @@
...
+        if (sharedPreferences?.getString(firstHero, "").equals("")) {
+            editor?.putString(firstHero, its)
+        } else if (sharedPreferences?.getString(firstHero, "").equals(its)) {
+            Toast.makeText(context, resources.getString(R.string.cantcompare), Toast.LENGTH_LONG)
+                .show()
+        } else {
+            editor?.putString(secondHero, its)
+            replaceFragment(CompareFragment.newInstance(), R.id.frameContainer)
+        }
+        editor?.apply()
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/hero_01/ui/champion/ChampionFragment.kt]: scope function editor

File: app/src/main/java/com/sun/hero_01/ui/compare/CompareFragment.kt [hunk 20]
```diff
@@ -1,13 +1,134 @@
...
+        exception?.let {
+            Toast.makeText(requireActivity(), it.message, Toast.LENGTH_SHORT).show()
+        }
+    }
+
+    override fun onStop() {
+        super.onStop()
+        val sharedPreferences: SharedPreferences? = context?.getSharedPreferences(PREF_COMPARE_HERO, Context.MODE_PRIVATE)
+        val editor = sharedPreferences?.edit()
+        editor?.let {
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/hero_01/ui/compare/CompareFragment.kt]:
```suggestion
        val editor = 
        sharedPreferences?.edit()?.let {
```

**Review by tuanlvt-1117 — APPROVED** [review]

**Pull request merged**